### PR TITLE
schannel: add support for NULL Encryption Cipher

### PR DESCRIPTION
**Pull request by essadiel**
This should add support for the eNULL cipher Suites.
NULL encryption algorithm is not directly supported by schannel
we need to set dwMinimumCipherStrength and
dwMaximumCipherStrength to -1
rational: NULL encryption is not often used but is sometime necessary,
for legal reasons.

### CODE REVIEW

File: lib/vtls/schannel.c [hunk 1]
```diff
@@ -337,6 +339,17 @@ set_ssl_ciphers(SCHANNEL_CRED *schannel_cred, char *ciphers)
     long alg = strtol(startCur, 0, 0);
     if(!alg)
       alg = get_alg_id_by_name(startCur);
+    if(alg == -1) {
```
Comment (bagder) [inline lib/vtls/schannel.c]: shouldn't this rather be `if(alg == eNULL)` for better readability ?

File: lib/vtls/schannel.c [hunk 2]
```diff
@@ -337,6 +339,17 @@ set_ssl_ciphers(SCHANNEL_CRED *schannel_cred, char *ciphers)
...
+        /* this is the eNULL case
+            to force Null encryption in schannel
+            we need to pass
+            dwMinimumCipherStrength & dwMaximumCipherStrength to -1
+            since this is a force case we ignore other algorithms
+        */
+         schannel_cred->dwMinimumCipherStrength = -1;
+         schannel_cred->dwMaximumCipherStrength = -1;
+         break;
```
Comment (bagder) [inline lib/vtls/schannel.c]: why break here, can't this be used in combination with another at times?

File: lib/vtls/schannel.c (hunk 1 above)
Comment (essadiel) [reply lib/vtls/schannel.c]: good point. i'll do the change as soon, as I can.

**Comment by bagder** [comment]
Ah can you also update the copyright year range to end in 2019?
~~~
./vtls/schannel.c:10:25: error: Copyright year out of date, should be 2019, is 2018 (COPYRIGHTYEAR)
  * Copyright (C) 2012 - 2018, Daniel Stenberg, <daniel@haxx.se>, et al.
                         ^

**Comment by nickzman** [comment]
Is this really a good idea? NULL cipher-suites are disabled in the Secure Transport code because they are not secure, and I didn't want to give users a false sense of security.

**Comment by mback2k** [comment]
I second the opinion of @nickzman. If this is added, I think there needs to be a big warning and I would also recommend to require a special compile-time flag to enable this feature. I understood that this can be a requirement for debugging or legal purposes, but I think we should make it hard to drop the security features.

**Comment by bagder** [comment]
I hear you both. I also question and doubt the "for legal reasons" argument based on the fact that few (other) TLS libraries allow this nowadays and surely only very few current HTTPS servers would agree to negotiate away the encryption like this?

**Comment by essadiel** [comment]
The reason I worked on that subject is because I was confronted to the problem of encryption for communicating with IMAPS protocol over the territory of china. SSL was still required of me for the handshake protocol and peer verification, but encryption of the message content was forbidden, by the state of China (who is known to block https website due to the encryption of communications)

**Comment by jay** [comment]
My opinion: invoker's choice. If they want to specify an insecure cipher why not, as long as it's not the default. how are we misleading the user, you mean the https:// url?

**Comment by mback2k** [comment]
@jay My reasoning for making it hard to drop the security is also that this avoids it sneaking into code examples were users just copy from, e.g. on stackoverflow or other places. It should be really obvious / clearly visible that the security is weakened if this option is active.

**Comment by bagder** [comment]
appveyor isn't happy:
~~~
2>schannel.c
2>.\vtls\schannel.c(349) : error C2220: warning treated as error - no 'object' file generated
2>.\vtls\schannel.c(349) : warning C4245: '=' : conversion from 'int' to 'DWORD', signed/unsigned mismatch
2>.\vtls\schannel.c(350) : warning C4245: '=' : conversion from 'int' to 'DWORD', signed/unsigned mismatch

**Comment by essadiel** [comment]
I don't really know what to do about the signed/unsigned warning: I understand where it come from, but on the other hand the documentation of schannel_cred explicitly said that it should be set to -1, casting -1 to DWORD remove the warning but do I do that and trust that windows know what they are doing?

also there is a spelling mismatch between VC and mingw  schannel_cred  (cipher vs cypher).

On the subject of should it be included, isn't it the same thing as using "rsa_null_sha" as cipher list with NSS?

**Comment by vszakats** [comment]
This looks like another kind of `-k` flag (and libcurl equivalent), which is already a huge and mostly hidden security risk, so I'm not sure adding another, similar option is a very good idea. As far as warnings go, `-k` always had one (at least in the command-line tool), yet it's widely used. Like discussed earlier with `-k`, maybe a "system-wide" (envvar) override to disable these insecure features may apply to this case as well, when implemented.

**Comment by bagder** [comment]
checksrc complains:
~~~
./vtls/schannel.c:349:82: warning: Longer than 79 columns (LONGLINE)
 /* this should be enought mingw and visual studio define __SCHANNEL_H__*/        
./vtls/schannel.c:349:81: warning: Trailing whitespace (TRAILINGSPACE)
 /* this should be enought mingw and visual studio define __SCHANNEL_H__*/        
./vtls/schannel.c:601:75: warning: Trailing whitespace (TRAILINGSPACE)
          infof(data,"schannel: Warning cipher used were manually set.\n"); 
                                                                           ^
./vtls/schannel.c:601:21: warning: comma without following space (COMMANOSPACE)
          infof(data,"schannel: Warning cipher used were manually set.\n"); 
                     ^

File: lib/vtls/schannel.c [hunk 3]
```diff
@@ -337,15 +340,32 @@ set_ssl_ciphers(SCHANNEL_CRED *schannel_cred, char *ciphers)
...
     if(!alg)
       alg = get_alg_id_by_name(startCur);
-    if(alg)
+    if(eNULL == alg) {
+        /* this is the eNULL case
+            to force Null encryption in schannel
+            we need to pass
+            dwMinimumCipherStrength & dwMaximumCipherStrength to -1
+            */
+/* this should be enought
```
Comment (jzakrzewski) [inline lib/vtls/schannel.c]: s/enought/enough -/

**Comment by essadiel** [comment]
I have to admit I have no clues as to why the last test fails: it is not even supposed to be affected by my change. 
On the subject of an environment variable to disable such an option, I didn't follow the "-k" flag discussion, so if someone could enlighten me. (my understanding is setting a #define to disable this part of the code, and have a variable at compilation to set that define, is that what it is?)

**Comment by bagder** [comment]
Please don't merge master into your branch, rebase your branch onto master instead and force-push so that your patch series is clean and easy to review. Then you can also squash the commits as you think is reasonable.

Also note that this now needs to be edited to fix a merge conflict.

I don't think the two test failures were your fault.

**Comment by bagder** [comment]
Is this PR abandoned now?

**Comment by jay** [comment]
I'm closing this due to lack of consensus.

**Pull request closed without merging**